csi/controller.go

```diff
@@ -115,19 +118,95 @@ func (s *OsdCsiServer) ControllerGetCapabilities(
 // ControllerPublishVolume is a CSI API implements the attachment of a volume
 // on to a node.
 func (s *OsdCsiServer) ControllerPublishVolume(
-	context.Context,
-	*csi.ControllerPublishVolumeRequest,
+	ctx context.Context,
+	req *csi.ControllerPublishVolumeRequest,
 ) (*csi.ControllerPublishVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "This request is not supported")
+	clogger.WithContext(ctx).Infof("ControllerPublishVolume request recived. VolumeID: %s", req.GetVolumeId())
+
+	vol, err := s.driverGetVolume(ctx, req.GetVolumeId())
+	if err != nil {
+		if s, ok := status.FromError(err); ok && s.Code() == codes.NotFound {
+			return nil, status.Error(codes.NotFound, "Volume not found")
+		}
+		return nil, err
+	}
+
+
+	// have the volumeID, submit a new grpc request to px driver to prepare
+	// volume to be published only if the volume propery is winshare.
+	conn, err := s.getRemoteConn(ctx)
+	if err != nil {
+		logrus.Errorf("failed to get remote connection: %v, continuing with local node instead", err)
+		conn, err = s.getConn()
+		if err != nil {
+			return nil, status.Errorf(
+				codes.Unavailable,
+				"Unable to connect to the SDK server: %v", err)
+		}
+	}
+
+	// TODO: Add checks for capability and readonly, etc.
```

Comment:
validation of volume capabilities missing.
see VolumeCreate API for example or others in this file pre-existing.
```
	if spec.IsPureVolume() {
		err = validateCreateVolumeCapabilitiesPure(req.GetVolumeCapabilities(), spec.GetProxySpec())
		if err != nil {
			return nil, err
		}
	} else {
		err = validateCreateVolumeCapabilities(req.GetVolumeCapabilities())
		if err != nil {
			return nil, err
		}
	}
```

Reply:
The new check, checks that capabilities exist, do we need to check that we have read/write/access permissions?

Reply:
All relevant checks needs to be confirmed. If the capabilities do not match, then it is best to stop servicing request much earlier.

Reply:
Yes please address this if it hasn't been already. If it has been addressed, comment here or mark resolved